File: vehicle-tracker-backend/asset-service/src/main/java/com/github/diogocerqueiralima/infrastructure/repositories/DeviceRepository.java
package com.github.diogocerqueiralima.infrastructure.repositories;

import com.github.diogocerqueiralima.infrastructure.entities.assets.DeviceEntity;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;
import java.util.UUID;

/**
 * Repository interface for managing {@link DeviceEntity} instances in the database.
 */
@Repository
public interface DeviceRepository extends JpaRepository<DeviceEntity, UUID> {

	/**
	 * Checks whether a device with the provided serial number exists.
	 *
	 * @param serialNumber the serial number to lookup.
	 * @return true when a record with the serial number exists.
	 */
	boolean existsBySerialNumber(String serialNumber);

	/**
	 * Checks whether a device with the provided IMEI exists.
	 *
	 * @param imei the IMEI to lookup.
	 * @return true when a record with the IMEI exists.
	 */
	boolean existsByImei(String imei);

	/**
	 * Finds a device by id and owner id.
	 *
	 * @param id device identifier.
	 * @param ownerId owner identifier.
	 * @return device when owned by the informed owner.
	 */
	Optional<DeviceEntity> findByIdAndOwnerId(UUID id, UUID ownerId);

	/**
	 * Retrieves a page of devices by owner id.
	 *
	 * @param ownerId owner identifier.
	 * @param pageable page request.
	 * @return paginated devices from the owner.
	 */
	Page<DeviceEntity> findAllByOwnerId(UUID ownerId, Pageable pageable);

}

File: vehicle-tracker-mobile/app/src/main/java/com/github/diogocerqueiralima/domain/repositories/DeviceRepository.kt
package com.github.diogocerqueiralima.domain.repositories

import com.github.diogocerqueiralima.domain.model.Device
import java.util.UUID

/**
 * Repository interface for managing Device entities.
 */
interface DeviceRepository {

    /**
     * Finds a Device by its unique identifier.
     *
     * @param id The unique identifier of the Device.
     * @return The Device if found, or null if no Device with the given ID exists.
     */
    suspend fun findById(id: UUID): Device?

    /**
     * @return A list of all Devices available in the repository.
     */
    suspend fun findAll(): List<Device>

    /**
     * Creates a new Device owned by the given user.
     *
     * @param serialNumber Manufacturer-assigned serial number of the device.
     * @param model Model name of the device.
     * @param manufacturer Manufacturer of the device.
     * @param imei International Mobile Equipment Identity of the device.
     * @param ownerId Identifier of the user who owns the device.
     * @return The created Device.
     */
    suspend fun create(serialNumber: String, model: String, manufacturer: String, imei: String, ownerId: UUID): Device

}
